python/graphstorm/sagemaker/sagemaker_partition.py

```diff
@@ -0,0 +1,371 @@
+    """
+    graph_data_s3: str
+    output_data_s3: str
+    num_parts: int
+    metadata_filename: str
+    partition_algorithm: str
+    skip_partitioning: bool
+    log_level: str
+
+def launch_build_dglgraph(
+        input_data_path,
+        partitions_dir,
+        ip_list,
+        output_path,
+        metadata_filename,
+        state_q):
+    """ Build DistDGL Graph
+
+    Parameters
+    ----------
+    input_data_path: str
+        Path to the input graph data
+    partitions_dir: str
+        Path to Per-node-type partid mappings.
+    ip_list: str
+        Path to a file storing ips of instances
+    output_path: str
+        Output Path
+    metadata_filename: str
+        The filename for the graph partitioning metadata file we'll use to determine data sources.
+    state_q: queue.Queue()
+        A queue used to return execution result (success or failure)
+    """
+    launch_cmd = ["python3", f"{DGL_TOOL_PATH}/dispatch_data.py",
+        "--in-dir", input_data_path,
+        "--metadata-filename", metadata_filename,
+        "--partitions-dir", partitions_dir,
+        "--ip-config", ip_list,
+        "--out-dir", output_path,
+        "--ssh-port", "22",
+        "--python-path", "/opt/conda/bin/python3",
+        "--log-level", logging.getLevelName(logging.root.getEffectiveLevel()),
+        "--save-orig-nids",
+        "--save-orig-eids"]
+
+    # launch postprocess task
+    thread = Thread(target=utils.run, args=(launch_cmd, state_q,), daemon=True)
+    thread.start()
+    # sleep for a while in case of ssh is rejected by peer due to busy connection
+    time.sleep(0.2)
+    return thread
+
+
+def download_graph(graph_data_s3, graph_config, world_size,
+        local_rank, local_path, sagemaker_session):
+    """ download graph structure data
+
+    Parameters
+    ----------
+    graph_data_s3: str
+        S3 uri storing the partitioned graph data
+    graph_config: dict
+        metadata config
+    world_size: int
+        Size of the cluster
+    local_rank: str
+        Path to store graph data
+    local_path: str
+        directory path under which the data will be downloaded
+    sagemaker_session: sagemaker.session.Session
+        sagemaker_session to run download
+
+    Return
+    ------
+    local_path: str
+        Local path to downloaded graph data
+    """
+
+    # download edge info
+    edges = graph_config["edges"]
+    for etype, edge_data in edges.items():
+        if local_rank == 0:
+            logging.info("Downloading edge structure for edge type '%s'", etype)
+        edge_file_list = edge_data["data"]
+        read_list = np.array_split(np.arange(len(edge_file_list)), world_size)
+        for i, efile in enumerate(edge_file_list):
+            # TODO: Only download round-robin if ParMETIS will run, skip otherwise
+            # Download files both in round robin and sequential assignment
+            if i % world_size == local_rank or i in read_list[local_rank]:
```

Comment:
Why both in round robin and sequential?
It is weird here.

Reply:
ParMETIS assumes round-robin assignment, while data_dispatch assume sequential. So we need to have both available locally. We could remove the round-robin for now since we only have random though.

Reply:
I'll have a follow-up PR that only downloads in the way that data_dispatch expects, would that cover it?